File: htmlvault/categorizer.py
"""
RETIRED — HTMLVault has moved to a Chrome extension as its only supported form.

This Python CLI (`htmlvault serve` / `htmlvault scan`) is kept in the repo
for historical reference only. It is no longer installed by pyproject.toml,
no longer published to PyPI, and no longer maintained. The logic here has
been reimplemented in `extension/` as MV3-compliant JavaScript.

If you are a non-technical user: please install the Chrome extension
(see README.md for instructions).

If you want to run the old CLI, revert to an earlier git tag (<= v0.1.0).
The entire original source is preserved below as a docstring so it can be
read, copied, or revived, but it will NOT execute as a module.
"""

_RETIRED_ORIGINAL_SOURCE = r"""
\"\"\"HTML type categorizer — detect slide/chart/dashboard/card/page.\"\"\"

import os
import re
from collections import OrderedDict
from typing import Optional

# Bounded LRU cache: filepath -> (mtime, category). Cap avoids unbounded growth
# in long-running serve sessions over directories with rotating outputs.
_CACHE_MAX = 1000
_cache: "OrderedDict[str, tuple]" = OrderedDict()


def categorize(filepath: str) -> str:
    \"\"\"Categorize an HTML file by analyzing its content.

    Results are cached by filepath + mtime — only re-reads if file changed.
    Returns one of: slide, chart, dashboard, card, page
    \"\"\"
    try:
        mtime = os.path.getmtime(filepath)
    except OSError:
        return "page"

    cached = _cache.get(filepath)
    if cached and cached[0] == mtime:
        _cache.move_to_end(filepath)
        return cached[1]

    try:
        with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
            content = f.read(200_000)  # read first 200KB max
    except (OSError, IOError):
        return "page"

    lower = content.lower()

    if _is_slide(lower, content):
        result = "slide"
    elif _is_chart(lower):
        result = "chart"
    elif _is_dashboard(lower):
        result = "dashboard"
    elif _is_card(lower, len(content)):
        result = "card"
    else:
        result = "page"

    _cache[filepath] = (mtime, result)
    _cache.move_to_end(filepath)
    if len(_cache) > _CACHE_MAX:
        _cache.popitem(last=False)
    return result


def _is_slide(lower: str, raw: str) -> bool:
    \"\"\"Detect presentation / slide deck.\"\"\"
    # Known slide libraries
    slide_libs = [
        "swiper", "reveal.js", "reveal.min", "impress.js",
        "slidev", "remark.js", "deck.js", "shower.js",
        "bespoke.js", "webslides",
    ]
    if any(lib in lower for lib in slide_libs):
        return True

    # Slide-like patterns
    slide_patterns = [
        "currentslide", "totalslides", "slideindex",
        "class=\"slide\"", "class='slide'",
        "data-slide", "slide-container",
        "prev-slide", "next-slide", "prevslide", "nextslide",
        "slide-nav", "slide-number", "slide-content",
    ]
    if any(p in lower for p in slide_patterns):
        return True

    # Multiple <section> tags with navigation
    section_count = lower.count("<section")
    if section_count >= 3:
        nav_hints = ["arrow", "prev", "next", "navigation", "indicator"]
        if any(h in lower for h in nav_hints):
            return True

    return False


def _is_chart(lower: str) -> bool:
    \"\"\"Detect chart / data visualization.\"\"\"
    chart_libs = [
        "echarts.init", "echarts.min", "new chart(",
        "chart.js", "chartjs", "d3.select", "d3.min",
        "highcharts", "apexcharts", "plotly",
        "google.visualization", "c3.generate",
        "recharts", "nivo", "visx",
    ]
    if any(lib in lower for lib in chart_libs):
        return True

    # Heavy SVG path usage (likely a chart)
    svg_paths = len(re.findall(r"<path\s+d=", lower))
    if svg_paths > 20:
        return True

    return False


def _is_dashboard(lower: str) -> bool:
    \"\"\"Detect dashboard / multi-panel layout.\"\"\"
    # Explicit dashboard keyword in meaningful context
    dash_patterns = [
        "dashboard", "data-panel", "panel-container",
        "stat-card", "metric-card", "kpi",
    ]
    dash_hits = sum(1 for p in dash_patterns if p in lower)
    if dash_hits >= 2:
        return True

    # Single strong signal
    if "dashboard" in lower:
        # Multiple grid/flex containers suggest multi-panel
        grid_count = lower.count("display:grid") + lower.count("display: grid")
        grid_count += lower.count("display:flex") + lower.count("display: flex")
        if grid_count >= 4:
            return True

    return False


def _is_card(lower: str, content_size: int) -> bool:
    \"\"\"Detect card / small fixed-canvas output.\"\"\"
    if content_size > 100_000:  # > 100KB is not a card
        return False

    # Fixed canvas dimensions
    fixed_patterns = [
        r"width:\s*1280px", r"width:\s*1080px", r"width:\s*720px",
        r"width:\s*800px", r"width:\s*640px",
        r"max-width:\s*1280px",
    ]
    has_fixed = any(re.search(p, lower) for p in fixed_patterns)

    # Small number of elements, card-like structure
    if has_fixed and lower.count("<div") < 50:
        return True

    return False
"""

File: extension/categorizer.js
// HTML type categorizer — detect slide/chart/dashboard/card/page.
// Ported 1:1 from htmlvault/categorizer.py.

// Cache: key (relpath + mtime) -> category
const _catCache = new Map();

/**
 * Categorize HTML content.
 * @param {string} content - HTML content (first 200KB)
 * @param {number} contentSize - total file size in bytes
 * @param {string} cacheKey - optional cache key (relpath + mtime)
 * @returns {string} one of: slide, chart, dashboard, card, page
 */
function categorize(content, contentSize, cacheKey) {
  if (cacheKey) {
    const cached = _catCache.get(cacheKey);
    if (cached) return cached;
  }

  const lower = content.toLowerCase();
  let result;

  if (_isSlide(lower, content)) {
    result = "slide";
  } else if (_isChart(lower)) {
    result = "chart";
  } else if (_isDashboard(lower)) {
    result = "dashboard";
  } else if (_isCard(lower, contentSize)) {
    result = "card";
  } else {
    result = "page";
  }

  if (cacheKey) _catCache.set(cacheKey, result);
  return result;
}

function _isSlide(lower, raw) {
  // Known slide libraries
  const slideLibs = [
    "swiper", "reveal.js", "reveal.min", "impress.js",
    "slidev", "remark.js", "deck.js", "shower.js",
    "bespoke.js", "webslides",
  ];
  if (slideLibs.some(lib => lower.includes(lib))) return true;

  // Slide-like patterns
  const slidePatterns = [
    "currentslide", "totalslides", "slideindex",
    'class="slide"', "class='slide'",
    "data-slide", "slide-container",
    "prev-slide", "next-slide", "prevslide", "nextslide",
    "slide-nav", "slide-number", "slide-content",
  ];
  if (slidePatterns.some(p => lower.includes(p))) return true;

  // Multiple <section> tags with navigation
  const sectionCount = (lower.match(/<section/g) || []).length;
  if (sectionCount >= 3) {
    const navHints = ["arrow", "prev", "next", "navigation", "indicator"];
    if (navHints.some(h => lower.includes(h))) return true;
  }

  return false;
}

function _isChart(lower) {
  const chartLibs = [
    "echarts.init", "echarts.min", "new chart(",
    "chart.js", "chartjs", "d3.select", "d3.min",
    "highcharts", "apexcharts", "plotly",
    "google.visualization", "c3.generate",
    "recharts", "nivo", "visx",
  ];
  if (chartLibs.some(lib => lower.includes(lib))) return true;

  // Heavy SVG path usage (likely a chart)
  const svgPaths = (lower.match(/<path\s+d=/g) || []).length;
  if (svgPaths > 20) return true;

  return false;
}

function _isDashboard(lower) {
  const dashPatterns = [
    "dashboard", "data-panel", "panel-container",
    "stat-card", "metric-card", "kpi",
  ];
  const dashHits = dashPatterns.filter(p => lower.includes(p)).length;
  if (dashHits >= 2) return true;

  if (lower.includes("dashboard")) {
    let gridCount = (lower.match(/display:\s*grid/g) || []).length;
    gridCount += (lower.match(/display:\s*flex/g) || []).length;
    if (gridCount >= 4) return true;
  }

  return false;
}

function _isCard(lower, contentSize) {
  if (contentSize > 100000) return false;

  const fixedPatterns = [
    /width:\s*1280px/, /width:\s*1080px/, /width:\s*720px/,
    /width:\s*800px/, /width:\s*640px/,
    /max-width:\s*1280px/,
  ];
  const hasFixed = fixedPatterns.some(p => p.test(lower));

  if (hasFixed && (lower.match(/<div/g) || []).length < 50) return true;

  return false;
}
